controller/src/main/java/com/esm/epam/controller/CertificateController.java

```diff
@@ -25,75 +26,98 @@
 
 import javax.validation.Valid;
 import javax.validation.constraints.Min;
-import java.util.ArrayList;
 import java.util.List;
+import java.util.Optional;
 
 import static com.esm.epam.validator.ControllerValidator.validateIntToBeUpdated;
 import static com.esm.epam.validator.ControllerValidator.validateSortValues;
+import static org.springframework.http.HttpStatus.OK;
 
 
 @RestController
 @RequestMapping("/certificates")
 @Validated
 public class CertificateController {
 
-    @Autowired
-    public CRUDService<Certificate> certificateService;
+    private final CertificateService certificateService;
+    private final HateoasBuilder<Certificate> hateoasBuilder;
 
-    @GetMapping
+    public CertificateController(CertificateService certificateService, HateoasBuilder<Certificate> hateoasBuilder) {
+        this.certificateService = certificateService;
+        this.hateoasBuilder = hateoasBuilder;
+    }
+
+    @GetMapping(params = {"page", "size"})
     @JsonView(View.UI.class)
-    public ResponseEntity<List<Certificate>> getCertificateList(@RequestParam(required = false) MultiValueMap<String, Object> params) throws ResourceNotFoundException, ControllerException {
-        List<Certificate> certificates = new ArrayList<>();
-        if (params.size() == 0) {
-            certificates = certificateService.getAll();
+    public ResponseEntity<List<Certificate>> getCertificateList(@RequestParam(required = false) MultiValueMap<String, Object> params, @RequestParam("page") @Min(0) int page, @RequestParam("size") @Min(1) int size) throws ResourceNotFoundException, ControllerException, ServiceException, DaoException {
+        List<Certificate> certificates;
+        if (params.size() == 2) {
+            certificates = certificateService.getAll(page, size);
         } else {
             validateSortValues(params);
-            certificates = certificateService.getFilteredList(params);
+            certificates = certificateService.getFilteredList(params, page, size);
+        }
+        for (Certificate certificate : certificates) {
+            hateoasBuilder.buildFullHateoas(certificate);
         }
         return new ResponseEntity<>(certificates, HttpStatus.OK);
     }
 
     @GetMapping("/{id}")
     @JsonView(View.UI.class)
-    public ResponseEntity<Certificate> getCertificate(@PathVariable("id") @Min(1L) Long id) throws ResourceNotFoundException, DaoException {
+    public ResponseEntity<Certificate> getCertificate(@PathVariable("id") @Min(1L) Long id) throws ResourceNotFoundException, DaoException, ControllerException, ServiceException {
```

Comment:
Controller should not throw any exceptions. All custom exceptions should be extended from RuntimeException not from Exception